File: src/nlp/sentiment_analyzer.py
"""
Sentiment Analyzer Module - Uses VADER for lightweight sentiment scoring.
Provides compound polarity scores and a convenience method to check
if sentiment requires escalation.
"""

from typing import Dict, Any
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer


class SentimentAnalyzer:
    """
    Wraps VADER SentimentIntensityAnalyzer.
    Provides compound sentiment scores and escalation threshold checking.
    """

    # Threshold below which sentiment is considered "escalation-worthy"
    DEFAULT_ESCALATION_THRESHOLD: float = -0.3

    def __init__(self, escalation_threshold: float = DEFAULT_ESCALATION_THRESHOLD) -> None:
        """
        Args:
            escalation_threshold: Compound score below which escalation is triggered.
        """
        self._analyzer = SentimentIntensityAnalyzer()
        self._escalation_threshold = escalation_threshold

    @property
    def escalation_threshold(self) -> float:
        return self._escalation_threshold

    def analyze(self, text: str) -> Dict[str, float]:
        """
        Analyze sentiment of a text string.

        Args:
            text: The user's message.

        Returns:
            Dictionary with keys: 'neg', 'neu', 'pos', 'compound'.
        """
        scores = self._analyzer.polarity_scores(text)
        return {
            "neg": round(scores["neg"], 4),
            "neu": round(scores["neu"], 4),
            "pos": round(scores["pos"], 4),
            "compound": round(scores["compound"], 4),
        }

    def needs_escalation(self, text: str) -> bool:
        """
        Check if sentiment is negative enough to warrant escalation.

        Args:
            text: The user's message.

        Returns:
            True if compound score is below the escalation threshold.
        """
        scores = self.analyze(text)
        return scores["compound"] < self._escalation_threshold

    def get_compound(self, text: str) -> float:
        """
        Convenience method to get the compound sentiment score directly.

        Args:
            text: The user's message.

        Returns:
            Compound polarity score between -1 (negative) and +1 (positive).
        """
        return self.analyze(text)["compound"]

File: lib/nlp/sentiment-analyzer.ts
/**
 * Sentiment Analyzer Module - TypeScript Port
 * Uses the 'sentiment' npm package (AFINN-based) for lightweight sentiment scoring.
 * Replaces the Python VADER SentimentIntensityAnalyzer.
 */

import Sentiment from "sentiment";

const sentimentAnalyzer = new Sentiment();

export interface SentimentResult {
  neg: number;
  neu: number;
  pos: number;
  compound: number;
}

// Escalation threshold: compound score below this triggers escalation
const DEFAULT_ESCALATION_THRESHOLD = -0.3;

/**
 * Analyze sentiment of a text string.
 * Returns normalized scores matching the Python VADER output format.
 */
export function analyzeSentiment(text: string): SentimentResult {
  const result = sentimentAnalyzer.analyze(text);

  // The 'sentiment' npm package gives: score, comparative, positive[], negative[]
  // We normalize to match VADER's neg/neu/pos/compound format
  const comparative = result.comparative;

  // Map comparative score (-5 to +5 range) to compound (-1 to +1)
  const compound = Math.max(-1, Math.min(1, comparative / 2));

  // Approximate neg/neu/pos from the word-level analysis
  const posWords = result.positive.length;
  const negWords = result.negative.length;
  const totalWords = result.tokens.length || 1;

  const neg = Math.round((negWords / totalWords) * 10000) / 10000;
  const pos = Math.round((posWords / totalWords) * 10000) / 10000;
  const neu = Math.round((1 - neg - pos) * 10000) / 10000;

  return {
    neg: Math.max(0, neg),
    neu: Math.max(0, Math.min(1, neu)),
    pos: Math.max(0, pos),
    compound: Math.round(compound * 10000) / 10000,
  };
}

/**
 * Check if sentiment is negative enough to warrant escalation.
 */
export function needsEscalation(
  text: string,
  threshold: number = DEFAULT_ESCALATION_THRESHOLD
): boolean {
  const scores = analyzeSentiment(text);
  return scores.compound < threshold;
}